### ISSUE TITLE
Need a way to specify test settings

### ISSUE THREAD

**Issue by Sleepwalk42** [nunit/nunit3-vs-adapter#49]
When running tests with the console runner, a number of settings can be specified, either through command line parameters, or through settings in a nunit project file. For most of these, there isn't an equivalent way to control the setting when running tests through the plugin.

One way to do this would be through the runsettings file, introduced with VS 2012. It looks like that file is intended for 3rd party extensions to be able to use. That file is at a solution level, so there may be multiple test assemblies present. In that case, it may be useful to specify some settings at either the solution level or the test assembly level (which would override a solution-level setting). Here are the settings that seem appropriate to me:
- application base
- private bin path
- process isolation
- domain
- x86
- timeout
- seed
- workers
- trace level
- shadow copy
- verbosity
- UseVsKeepEngineRunning

In particular, I'd like to enable scenarios where the assemblies containing unit tests and the assemblies being tested (which need to be referenced by the assemblies containing the tests) are output into separate parallel folders. Normally this would present a problem in loading the assemblies being tested, without copying them to over to the folder containing the tests, or some other hack. It would be much cleaner to set the application base to a higher level folder, and use the private bin path to specify the probing path (which needs to be inside the application base folder) for assembly resolution. It should be possible to specify the application base relative to the solution directory, or maybe the directory containing the runsettings file. That way, it'll work correctly regardless of the full path to the source tree.

It would also be good if settings specified in the runsettings file for the solution could be automatically used when passing the solution to the console runner. That way, settings only need to be changed in one place, which makes it easier to get consistent test behavior in both the adapter and the console runner.

**Comment by CharliePoole** [comment]
Thanks for creating this issue. It's a good summary of settings that one might want to control on a run-by-run basis.

Can you give an example of how we might express solution- versus assembly-level settings?

**Comment by CharliePoole** [comment]
Looking at the layout of the RunSettings file, I came up with the following equivalences...

RunConfiguration Element
  ResultsDirectory = use to set NUnit WorkDirectory
  TargetPlatform = if X86, set NUnit to run X86 [No way to tell NUnit to use X64, it's the default.]
  TargetFrameworkVersion = translate VS values to NUnit framework option values.

NUnit Element (new settings for NUnit)
  ApplicationBase
  PrivateBinpath (or ProbingPath?)
  DefaultTimeOut
  Seed
  Workers
  InternalTraceLevel
  ShadowCopy
  Verbosity
  UseVsKeepEngineRunning

We could also support equivalent elements to the RunConfiguration elements under NUnit. However, my first guess is that it makes sense to use RunConfiguration in case of running NUnit tests along with some other kind of tests.

The process and domain options are not currently supported under the adapter at all. We always run in the process created by VS and create a separate domain for each test assembly. We could add the relevant settings if this ever changes.

Comments?

**Comment by CharliePoole** [comment]
@OsirisTerje I'd particularly like to hear from you about this design before I implement it!

**Comment by OsirisTerje** [comment]
It is a good idea to read off the settings from the runsettings file.  We have discussed earlier some more of this, and we should also add the [New] runparameters.  
The x86 is controlled by VSTest, there are two runners that might be invoked, one x86 and one x64, trying to change what they start is probably not wise.   The same with the framework, that should be read off the Projects targets,  but I thought both of these were handled today ?
Verbosity and UseVSKeepEnginerunning is used by the adapter, Shadowcopy is present today and is passed down, but having it in runsettings might be a better place.  
Also, if these settings could be adjusted from a Tools/Options/NUnit page it would be nice, writing down to a runsettings file. 
Process/Domain should be seen in combination with the Parallell option we're working on.  With that in, we will have different process/domain options, so some Control over those might be in order, we just have to make sure we don't set opposing options (Parallell without Process).

**Comment by CharliePoole** [comment]
I see that VS provides the RunConfiguration element to set the framework target and whether to use X86 or X64. As far as I know, the defaults work fine, but if somebody uses those elements in a runsettings file, I assume that VS will honor them. I agree that there is no need for us to act on them.

In the case of ResultDirectory, which is also implemented by VS, I think we should use the same directory for our output, if the user has specified it. That's a small thing.

I think an Options page could be a future enhancement we can undertake. It's not entirely clear to me whether the usual Options page is intended to apply to a single run or to set default behavior. In this issue, we want to deal with the individual run.

Process will be something we need to set if/when we implement Parallel. It seems wrong to implement it now, even though it might save a little time later. I'm not sure what you mean about setting Parallel without Process... Parallel is simply a value we allow for Process, similar to Single and Multiple.

**Comment by Sleepwalk42** [comment]
@CharliePoole Here's a quick mockup of what I was thinking for specifying settings at different levels, that would enable the test/development setup described in [this post.](https://groups.google.com/d/topic/nunit-discuss/bj1ggukWR6s/discussion) It also sets things at the configuration level (probably of main importance to paths, since builds typically output to different locations for different configs). I was originally thinking that the block in the runsettings file could have the same syntax as an nunit project file (which might make it possible to reuse some of the code that handles it), but I don't know if that would actually work well in practice.

```XML
<NUnit>
<!-- appbase could be given as an absolute path, or relative to the directory containing the runsettings file. In this example, the runsettings file is at C:\SolutionDir.  -->

<!-- This sets the application base at the solution level to C:\SolutionDir -->
<AppBase>.</AppBase>
<Config name="Debug" binpath="output\Debug\Test;output\Debug\Program"/>

<!-- This sets the application base for the release configuration to C:\SolutionDir\output\Release -->
 <Config name="Release" appbase="output\Release" binpath="Test;Program">
    <!-- For the assembly at C:\SolutionDir\output\Release\Test\TestAssembly.dll, in the Release configuration, set the timeout to 60 seconds. -->
    <assembly path="Test\TestAssembly.dll" timeout="60000"/>
 </Config>
</NUnit>
```

I think the idea I'm trying to convey is that any settings made at the top level (outside a configuration element, or an assembly element) are applied to the whole solution. Setting them at a lower level would override the setting from a higher level, which may just be at the default value. It might also be useful to be able to have an assembly element outside a config element, meaning you're specifying settings for the assembly regardless of configuration. For the settings themselves, you could define them to be attributes or elements. Should be able to make it clear either way.

Unlike nunit project files, not all of the test assemblies need to be specified, since the interface for the test adapter passes those as parameters. Just need to specify assemblies that would have settings that you want to specify.

**Comment by CharliePoole** [comment]
This is quite an advanced idea. :-)

How do you propose that we figure out what configuration is being run? Remember, we only have the path to the assembly as passed from VisualStudio. A further question would be how to apply this when asked to run selected tests.

That's not to say it's a bad idea. I think we'll start of fairly simply and see where it takes us.

**Comment by CharliePoole** [comment]
Another possibility is to define an element in the runsettings file that points to an NUnit project. Hmmm.

**Comment by Sleepwalk42** [comment]
I wasn't sure what all information Visual Studio is providing. I couldn't find any API documentation for the interfaces. As far as I know, there's always an active configuration for an open solution in VS. Some quick googling turned up [this](http://stackoverflow.com/q/11868290) and [this](http://www.mztools.com/articles/2013/MZ2013007.aspx) for ways of getting the current solution configuration in an add-in. If running the solution via the console runner, you'd have to specify the config in a parameter, and have a default if one wasn't provided (seems like that's the way nunit project files work).

**Comment by Sleepwalk42** [comment]
The configuration stuff falls apart if there isn't any way to determine the solution's currently active configuration in the test adapter extension.

As far as how to apply the configuration when asked to run selected tests, the logic for applying the settings should work exactly the same as when running everything, you're just running fewer tests. For any given test that you're running, I'd expect the order of priority to resolve each setting to be as follows:
1. Setting specified for the assembly containing the test, inside the current configuration
2. Setting specified for the assembly containing the test, outside any configuration
3. Setting specified for the current configuration
4. Setting specified outside any assembly or configuration
5. Default value for the setting

You could simplify it a bit by making the assembly always be inside a configuration, and not having the option of specifying settings at the solution level (so they would always be either at the configuration level or the assembly level). That would remove items 2 and 4 from the list above, but wouldn't remove any flexibility of what you could do. It would just force users to specify solution-level and assembly-level settings multiple times (once for each configuration).

**Comment by CharliePoole** [comment]
Regarding lack of API documentation, welcome to the club. It has been lots of fun writing and maintaining the adapter in it's absence, especially after the initial enthusiasm on the part of the Microsoft guys died down!

Yes, it's easy to get the configuration from an addin, but the adapter isn't a VS addin. Rather it's an extension to a Microsoft-written extension. We could write an addin that was bundled with the adapter, but that's more work than I want to do for this purpose. If I were writing an extension I think I would tend toward replacing the entire functionality of the TestWindow for NUnit only.

However, it's most common for desired settings to "stick" with assemblies at a given path location. So we could have different values for bin\debug\myassembly.dll and bin\Release\myassembly.dll. I'm not sure that's really needed, however. NUnit does a fine job of deducing that probing path provided you give it an appbase to start from. Are there other settings you might want to have set differently based on config?

In any case, my inclination is to get the most basic implementation up and running and then see if other needs arise. The easiest and most bug-free code is the code you never wrote. :-)

**Comment by Sleepwalk42** [comment]
Those are the only settings that I'm interested in. I think it would be really odd to want to have most of the other settings differ between configurations, but then again it wouldn't surprise me if there were developers out there that did odd things differently between configurations. I'd like to be able to streamline things, so that if I have a bunch of test assemblies in output\Release\Test, they can all resolve assembly references from output\Release\Program, and similarly from output\Debug\Test to output\Debug\Program. Ideally I'd be able to set that up so it happened automatically, so that if someone added a new test assembly to the solution, it would just work, without needing to add settings for the debug and release versions of the assembly to the runsettings file. 

I think that automatic behavior would require getting the current configuration. If getting the current configuration is simply not possible inside a test adapter, then so be it.

**Comment by CharliePoole** [comment]
I guess you are right for your configuration. If the Appbase is set to the solution directory and the test assembly is at ./tests/testassembly, NUnit will automatically set the probing path to tests. However, you need it to pick up the class under test as well, so you want the path to be software;tests.

The only simple way to do this would be to use a fixed appbase and probing path for all tests in each config and to have a separate settings file for debug and release.

Again, we'll learn more when somebody picks this up to work on.

**Comment by FermJacob** [comment]
+1 for this.  I want to move from using VSTest to NUnit because of how VSTest parallelizes its tests.  I am using VSTS and the .runsettings with the example below to have default parameters that can be overrode at runtime. 

<TestRunParameters>
	<Parameter name="webAppUrl" value="https://google.com" />
	<Parameter name="Username" value="username" />
	<Parameter name="Password" value="password" />
</TestRunParameters>

**Comment by CharliePoole** [comment]
@FermJacob There's no "example below." :-)

**Comment by FermJacob** [comment]
@CharliePoole added comments.  Let me know if you would like more details and sample implementation

**Comment by CharliePoole** [comment]
@FermJacob Thanks for the info. Up to now, this issue has been about providing a way to pass settings to NUnit itself, not to the user test.

NUnit itself, on which the adapter depends, does not have a way to pass arguments to the test, although this has been requested in the past. See issue nunit/nunit#720.

If we once give NUnit this feature, then it would be a matter of providing some way for it to be passed in by the adapter.

As a matter of interest, how do you access settings in your code when using VSTS?

**Comment by FermJacob** [comment]
Sure thing!

With the new version of Visual Studio 2015 (Update 1?), they say to use the .runsettings file: [MSDN](https://msdn.microsoft.com/en-us/library/jj635153.aspx).  They state inside the ForcedLegacyMode "We recommend that you consider refactoring your tests to allow you to use the newer adapter."

Here is the original blog post about the TestRunParameters: [Blog](https://blogs.msdn.microsoft.com/visualstudioalm/2015/09/04/supplying-run-time-parameters-to-tests/) .  It gives a good example on how to implement it and override the parameter with VSTest at runtime with the VSTS builds.  Here is the shared method I am using:

```csharp
public static string GetString(TestContext context, string property)
        {
            if (context.Properties[property] == null)
            {
                throw new ArgumentException("Property does not exist or does not have a value");
            }
            else
            {
                return context.Properties[property].ToString();
            }
        }
```

I am using this with Selenium tests to override which browser to use and the URL of the web application at runtime for mutliple environments.  Like I said above, I am looking to switch to NUnit because VSTest runs tests in parallel at the assembly not the class/test(I know test hasn't been implemented) like NUnit to speed up execution time.  

I hope this answers your question.  Let me know if you need more info.

**Comment by CharliePoole** [comment]
Thanks for the info, which will be helpful. In the NUnit issue nunit/nunit#720, we will have to come up with a way to access any properties. I'd been planning on using TestContext, which is an NUnit class, but with no relation to the VSTS class of the same name. Still, it would probably make it comfortable for folks converting to NUnit.

Up to now this issue has been about implementing an `<NUnit>` element to the RunSettings file, which is a big enough thing in itself. It would be handy to have a separate issue around the use of `<TestRunParameters>` as well. Would you mind creating one and pointing to this discussion? While this issue (#49) stands on it's own, the new one will be blocked by nunit/nunit#720 until it is resolved.
